MdeModulePkg/Include/Library/MemoryBucketLib.h

```diff
@@ -0,0 +1,130 @@
+/** @file
+
+Library for defining memory buckets.  Can be referenced externally.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __MEMORY_BUCKET_LIB_H__
+#define __MEMORY_BUCKET_LIB_H__
+
+#include <Uefi.h>
+#include <Base.h>
+#include <Guid/MemoryBucketInformation.h>
+
+/**
+  Internal function to build a HOB for the memory allocation.
+  It will search and reuse the unused(freed) memory allocation HOB,
+  or build memory allocation HOB normally if no unused(freed) memory allocation HOB found.
+
+  @param[in] BaseAddress        The 64 bit physical address of the memory.
+  @param[in] Length             The length of the memory allocation in bytes.
+  @param[in] MemoryType         The type of memory allocated by this HOB.
+
+**/
+VOID
+InternalBuildRuntimeMemoryAllocationHob (
+  IN EFI_PHYSICAL_ADDRESS  BaseAddress,
+  IN UINT64                Length,
+  IN EFI_MEMORY_TYPE       MemoryType
+  );
+

```

Comment:
Generally, internal functions should not be part a library external interface.

Can you give a bit more info about why some of these functions are here?

Reply:
These functions are taken directly from PEI memory services where these are used for internal hob memory allocations.  I end up needing them for the same reason.  I can move these functions to MemoryBucketHelperLib to not make them forward facing.  Another option is that I can consolidate MemoryBucketLib with MemoryServices if that makes more sense.  I didn't want to change one of the core services too much but it might be a better path.